### PR TITLE
Reproduce Figure 2 with GPT-2

### PR DESCRIPTION
**Pull request by shaheenahmedc**
In this MR I:
- Add `scripts/generate_next_token_prediction_data.py` to get the same data that the Patchscope authors use for Figure 2. 
- Update `scripts/token_identity.py` to use the output of `scripts/generate_next_token_prediction_data.py`.
- Add `scripts/replicate_figure_2.py` to output surprisal plots for `ClassicLogitLens`, `PatchscopeLogitLens` and `TokenIdentity` patchscopes, as seen in Figure 2 of the patchscopes paper. I wanted to do this for gpt-j but ran out of memory. 
- Make a small edit to `obvs/lenses.py`, due to a quirk with gpt-j. It's tokenizer has vocab_size of 50257, but its embedding d_vocab is 50400. I might be wrong here, but I could've sworn these were meant to match. See [here](https://huggingface.co/EleutherAI/gpt-j-6b) for details. 
- Add (repeated!) surprisal code to each of `ClassicLogitLens` and `PatchscopeLogitLens`. I initially wanted to re-factor these out, but got bogged down with the TokenIdentity `run_and_compute()` method using batched surprisal calculations from `obvs/metrics.py`. To do: add this as an issue.
- Add some packages to `pyproject.toml`, mainly around jupyter, and zstandard to make `obvs/generate_next_token_prediction_data.py` work. 
- Github actions was causing headaches between poetry.lock mismatch, even though I'd ran poetry update. I think it was because github actions and pre-commit still used 3.11. Switched these to 3.10.13. Note: this didn't fix the issue, but it seems worth keeping these changes. `poetry cache clear PyPI --all` fixed it. It then broke on requiring a pre-processed file from `scripts/generate_next_token_prediction_data.py`. Fixed this by getting pytest in github actions main.yaml to ignore the scripts folder. Lmk if you disagree with this solution, moderately uncertain about it. Doesn't seem sensible to push pre-processed data from scripts to github? But is ignoring scripts folder in pytest overkill?

### CODE REVIEW

File: obvs/lenses.py [hunk 1]
```diff
@@ -418,3 +450,34 @@ def run(self, substring: str, layers: list[int]):
         self.data["logits"] = self.data["logits"].detach()
         self.data["substring_tokens"] = substring_tokens
         self.data["layers"] = layers
+
+    def compute_surprisal_at_position(self):
```
Comment (tvhong) [inline obvs/lenses.py:454]: Can we dedupe the code by implementing this method in `BaseLogitLens`?

File: scripts/replicate_figure_2.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+layers = list(range(0, num_layers))
+# Define the number of sentences to run
+test_amount = 100
+# Initialize the arrays to store the surprisal values
+surprisals_log_classic_ll = np.zeros((test_amount, num_layers))
+surprisals_log_patchscope_ll = np.zeros((test_amount, num_layers))
+surprisals_log_ti = np.zeros((test_amount, num_layers))
+warnings.filterwarnings("ignore")  # Ignore all warnings
+
+# Loop through the data
```
Comment (tvhong) [inline scripts/replicate_figure_2.py:55]: In general, we should avoid commenting on **what** the code is doing. Instead, comments should be explaining **why** we're doing something.

This and the next few comments are repeats of what each line does. They add noise and make the code harder to read.

File: scripts/test_logit_lens_surprisal.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
+from __future__ import annotations
```
Comment (tvhong) [inline scripts/test_logit_lens_surprisal.py:1]: Is this file a duplicate of `replicate_figure_2.py`? If so, can we remove it?

File: obvs/lenses.py (hunk 1 above)
Comment (shaheenahmedc) [reply obvs/lenses.py]: Ah of course, thanks! Done

File: obvs/lenses.py [hunk 4]
```diff
@@ -365,6 +365,37 @@ def run(self, position: int):
         self.data["substring_tokens"] = self.substring_tokens
         self.data["layers"] = self.layers
 
+    def compute_surprisal_at_position(self):
+        """Compute the surprisal for a specific position across all layers.
+
+        Args:
+            target_position (int): The position in the token sequence for which to compute surprisal.
```
[the remark this answers is not in the document]
Comment (shaheenahmedc) [reply obvs/lenses.py]: Removed in base logit lens refactoring

File: scripts/replicate_figure_2.py (hunk 2 above)
Comment (shaheenahmedc) [reply scripts/replicate_figure_2.py]: Thanks, very helpful. Removed unnecessary comments

File: scripts/test_logit_lens_surprisal.py (hunk 3 above)
Comment (shaheenahmedc) [reply scripts/test_logit_lens_surprisal.py]: Yep, was a legacy file with just the logit lenses, and not the token identity lens. Removed

**Review by tvhong — APPROVED** [review]
LGTM

**Pull request merged**